File: src/com/turlington/JobListing.java
package com.turlington;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

/**
 * Information about a job.
 * Created by Mitchell on 7/18/2016.
 */
public class JobListing {
    private final String title;
    private String url = null;
    private Map<String, String> jobProperties = new HashMap<>();
    private String emailSubject;

    JobListing(String title) {
        this.title = title;
    }

    void addProperty(String name, String value) {
        jobProperties.put(name, value);
    }

    private Map<String, String> getProperties() {
        return jobProperties;
    }

    private Optional<String> getUrl() {
        if (url != null) {
            return Optional.of(url);
        }
        return Optional.empty();
    }

    void setUrl(String url) {
        this.url = url;
    }

    String getTitle() {
        return title;
    }

    String getEmailText() {
        StringBuilder text = new StringBuilder("Title: " + getTitle());
        if (getUrl().isPresent()) {
            text.append("\nLink: ").append(getUrl().get());
        }
        for (Map.Entry<String, String> entry : getProperties().entrySet()) {
            text.append("\n").append(entry.getKey()).append(": ").append(entry.getValue());
        }
        return text.toString();
    }

    void setEmailSubject(String emailSubject) {
        this.emailSubject = emailSubject;
    }

    String getEmailSubject() {
        return emailSubject;
    }
}

File: src/main/kotlin/com/turlington/JobListing.kt
package com.turlington

import java.util.*

/**
 * Information about a job.
 * Created by Valerie on 7/18/2016.
 */
class JobListing internal constructor(internal val title: String, private var url: String? = null,
                                      internal var emailSubject: String = "{Default Subject}") {
    private val jobProperties = HashMap<String, String>()

    internal fun addProperty(name: String, value: String) {
        jobProperties.put(name, value)
    }

    private val properties: Map<String, String>
        get() = jobProperties

    internal fun setUrl(url: String) {
        this.url = url
    }

    internal val emailText: String
        get() {
            val text = StringBuilder("Title: $title")
            if (url != null) {
                text.append("<br>Link: ").append(url)
            }
            for ((key, value) in properties) {
                text.append("<br>").append(key).append(": ").append(value)
            }
            return text.toString()
        }
}
